File: src/rpc/weapon.rs
use stubs::weapon::v0::weapon_service_server::WeaponService;
use stubs::*;
use tonic::{Request, Response, Status};

use super::MissionRpc;

#[tonic::async_trait]
impl WeaponService for MissionRpc {
    async fn get_launcher(
        &self,
        request: Request<weapon::v0::GetLauncherRequest>,
    ) -> Result<Response<weapon::v0::GetLauncherResponse>, Status> {
        let res = self.request("weaponGetLauncher", request).await?;
        Ok(Response::new(res))
    }

    async fn get_target(
        &self,
        request: Request<weapon::v0::GetTargetRequest>,
    ) -> Result<Response<weapon::v0::GetTargetResponse>, Status> {
        let res = self.request("weaponGetTarget", request).await?;
        Ok(Response::new(res))
    }

    async fn get_category(
        &self,
        request: Request<weapon::v0::GetCategoryRequest>,
    ) -> Result<Response<weapon::v0::GetCategoryResponse>, Status> {
        let res = self.request("weaponGetCategory", request).await?;
        Ok(Response::new(res))
    }

    async fn get_desc(
        &self,
        request: Request<weapon::v0::GetDescRequest>,
    ) -> Result<Response<weapon::v0::GetDescResponse>, Status> {
        let res = self.request("weaponGetDesc", request).await?;
        Ok(Response::new(res))
    }

    async fn get_position(
        &self,
        request: Request<weapon::v0::GetPositionRequest>,
    ) -> Result<Response<weapon::v0::GetPositionResponse>, Status> {
        let res = self.request("weaponGetPosition", request).await?;
        Ok(Response::new(res))
    }

    async fn get_velocity(
        &self,
        request: Request<weapon::v0::GetVelocityRequest>,
    ) -> Result<Response<weapon::v0::GetVelocityResponse>, Status> {
        let res = self.request("weaponGetVelocity", request).await?;
        Ok(Response::new(res))
    }

    async fn in_air(
        &self,
        request: Request<weapon::v0::InAirRequest>,
    ) -> Result<Response<weapon::v0::InAirResponse>, Status> {
        let res = self.request("weaponInAir", request).await?;
        Ok(Response::new(res))
    }

    async fn is_exist(
        &self,
        request: Request<weapon::v0::IsExistRequest>,
    ) -> Result<Response<weapon::v0::IsExistResponse>, Status> {
        let res = self.request("weaponIsExist", request).await?;
        Ok(Response::new(res))
    }

    async fn destroy(
        &self,
        request: Request<weapon::v0::DestroyRequest>,
    ) -> Result<Response<weapon::v0::DestroyResponse>, Status> {
        let res = self.request("weaponDestroy", request).await?;
        Ok(Response::new(res))
    }

    async fn get_coalition(
        &self,
        request: Request<weapon::v0::GetCoalitionRequest>,
    ) -> Result<Response<weapon::v0::GetCoalitionResponse>, Status> {
        let res = self.request("weaponGetCoalition", request).await?;
        Ok(Response::new(res))
    }

    async fn get_country(
        &self,
        request: Request<weapon::v0::GetCountryRequest>,
    ) -> Result<Response<weapon::v0::GetCountryResponse>, Status> {
        let res = self.request("weaponGetCountry", request).await?;
        Ok(Response::new(res))
    }

    async fn get_name(
        &self,
        request: Request<weapon::v0::GetNameRequest>,
    ) -> Result<Response<weapon::v0::GetNameResponse>, Status> {
        let res = self.request("weaponGetName", request).await?;
        Ok(Response::new(res))
    }

    async fn get_type_name(
        &self,
        request: Request<weapon::v0::GetTypeNameRequest>,
    ) -> Result<Response<weapon::v0::GetTypeNameResponse>, Status> {
        let res = self.request("weaponGetTypeName", request).await?;
        Ok(Response::new(res))
    }

    async fn get_point(
        &self,
        request: Request<weapon::v0::GetPointRequest>,
    ) -> Result<Response<weapon::v0::GetPointResponse>, Status> {
        let res = self.request("weaponGetPoint", request).await?;
        Ok(Response::new(res))
    }
}

File: lua/DCS-gRPC/methods/weapon.lua
GRPC.ActiveWeapons = {}

GRPC.methods.weaponGetLauncher = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.errorNotFound("weapon not found") end
  local launcher = w:getLauncher()
  if not launcher then return GRPC.success({}) end
  return GRPC.success({ launcherName = launcher:getName() })
end

GRPC.methods.weaponGetTarget = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.errorNotFound("weapon not found") end
  local target = w:getTarget()
  if not target then return GRPC.success({}) end
  return GRPC.success({ targetName = target:getName() })
end

GRPC.methods.weaponGetCategory = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.errorNotFound("weapon not found") end
  -- gRPC enum for object category starts at 1
  return GRPC.success({ category = w:getCategory() + 1 })
end

GRPC.methods.weaponGetDesc = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.errorNotFound("weapon not found") end
  -- DCS getDesc returns a table. We encode it to JSON string.
  return GRPC.success({ descJson = net.lua2json(w:getDesc()) })
end

GRPC.methods.weaponGetPosition = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.errorNotFound("weapon not found") end
  local pos = w:getPosition()
  if not pos then return GRPC.errorNotFound("could not get position") end

  local lat, lon, alt = coord.LOtoLL(pos.p)
  return GRPC.success({
    position = {
      position = { lat = lat, lon = lon, alt = alt },
      forward = pos.x,
      up = pos.y,
      right = pos.z
    }
  })
end

GRPC.methods.weaponGetVelocity = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.errorNotFound("weapon not found") end
  local v = w:getVelocity()
  if not v then return GRPC.errorNotFound("could not get velocity") end
  return GRPC.success({
    velocity = { x = v.x, y = v.y, z = v.z }
  })
end

GRPC.methods.weaponInAir = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.errorNotFound("weapon not found") end
  return GRPC.success({ inAir = w:inAir() })
end

GRPC.methods.weaponIsExist = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.success({ isExist = false }) end
  return GRPC.success({ isExist = w:isExist() })
end

GRPC.methods.weaponDestroy = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.errorNotFound("weapon not found") end
  w:destroy()
  return GRPC.success({})
end

GRPC.methods.weaponGetCoalition = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.errorNotFound("weapon not found") end
  -- gRPC enum for Coalition starts at 1, DCS starts at 0 (Neutral = 0)
  return GRPC.success({ coalition = w:getCoalition() + 1 })
end

GRPC.methods.weaponGetCountry = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.errorNotFound("weapon not found") end
  return GRPC.success({ country = w:getCountry() })
end

GRPC.methods.weaponGetName = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.errorNotFound("weapon not found") end
  return GRPC.success({ name = w:getName() })
end

GRPC.methods.weaponGetTypeName = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.errorNotFound("weapon not found") end
  return GRPC.success({ typeName = w:getTypeName() })
end

GRPC.methods.weaponGetPoint = function(params)
  local w = GRPC.ActiveWeapons and GRPC.ActiveWeapons[params.name]
  if not w then return GRPC.errorNotFound("weapon not found") end
  local p = w:getPoint()
  if not p then return GRPC.errorNotFound("could not get point") end
  local lat, lon, alt = coord.LOtoLL(p)
  return GRPC.success({
    point = { lat = lat, lon = lon, alt = alt }
  })
end
